File: base/code/quicksort/quicksort.erl

%%----------------------------------------------------------------------
%% qsort/1
%%
%% Tail recursive quicksort
%%----------------------------------------------------------------------
qsort([]) ->
    [];
qsort([H | _] = List) ->
    {L, E, G} = partition(H, List, {[], [], []}),
    qsort(L) ++ E ++ qsort(G).

partition(_, [], {L, E, G}) ->
    {L, E, G};
partition(Pivot, [H | List], {L, E, G}) when Pivot > H ->
    partition(Pivot, List, {[H | L], E, G});
partition(Pivot, [H | List], {L, E, G}) when Pivot < H ->
    partition(Pivot, List, {L, E, [H | G]});
partition(Pivot, [H | List], {L, E, G}) when Pivot == H ->
    partition(Pivot, List, {L, [H | E], G}).

File: base/code/quicksort/quicksort.jl
function quicksort!(a,lo,hi)
    i, j = lo, hi
    while i < hi
        pivot = a[(lo+hi)>>>1]
        while i <= j
            while a[i] < pivot; i = i+1; end
            while a[j] > pivot; j = j-1; end
            if i <= j
                a[i], a[j] = a[j], a[i]
                i, j = i+1, j-1
            end
        end
        if lo < j; quicksort!(a,lo,j); end
        lo, j = i, hi
    end
    return a
end
